client_plugin/drivers/shared/kvstore/kvstore.go

```diff
@@ -0,0 +1,89 @@
+
+package kvstore
+
+// VolStatus: Datatype for keeping status of a shared volume
+type VolStatus string
+
+/*
+   Constants:
+   VolStateCreating:     Shared volume is being created. Not ready to be mounted.
+   VolStateReady:        Shared volume is ready to be mounted but
+                         no Samba service running right now.
+   VolStateMounted:      Samba service already running. Volume mounted
+                         on at least one host VM.
+   VolStateMounting:     Volume is being mounted, File sharing service is being
+                         started for this volume.
+   VolStateUnmounting:   Volume is being unmounted, File sharing service is being
+                         stopped for this volume.
+   VolStateDeleting:     Shared volume is being deleted. Not ready to be mounted.
+   VolStateError:        Shared volume in error state.
+
+   VolPrefixState:       Each volume has three metadata keys. Each such
+                         key terminates in the name of the volume, but
+                         is preceded by a prefix. This is the prefix
+                         for "State" key
+   VolPrefixGRef:        The prefix for GRef key (Global refcount)
+   VolPrefixInfo:        The prefix for info key. This key holds all
+                         other metadata fields squashed into one
+
+   VolumeDoesNotExistError:    Error indicating that there is no such volume
+*/
+const (
+	VolStateCreating        VolStatus = "Creating"
+	VolStateReady           VolStatus = "Ready"
+	VolStateMounted         VolStatus = "Mounted"
+	VolStateMounting        VolStatus = "Mounting"
+	VolStateUnmounting      VolStatus = "Unmounting"
+	VolStateDeleting        VolStatus = "Deleting"
+	VolStateError           VolStatus = "Error"
+	VolPrefixState                    = "SVOLS_stat_"
+	VolPrefixGRef                     = "SVOLS_gref_"
+	VolPrefixInfo                     = "SVOLS_info_"
+	VolumeDoesNotExistError           = "No such volume"
+)
+
+// KvPair : Key Value pair holder
+type KvPair struct {
+	Key   string
+	Value string
+}
+
+// KvStore is the interface for VolumeDriver to access a plugin-level KV store
+type KvStore interface {
+	// WriteMetaData - Update or Create volume metadata in KV store
+	WriteMetaData(entries []KvPair) error
+
+	// ReadMetaData - Read volume metadata in KV store
+	ReadMetaData(keys []string) ([]KvPair, error)
+
+	// DeleteMetaData - Delete volume metadata in KV store
+	DeleteMetaData(name string) error
+
+	// CompareAndPut - Compare the value of key with oldVal, if equal, replace with newVal
+	CompareAndPut(key string, oldVal string, newVal string) bool
+
+	// CompareAndPutStateOrBusywait - Compare the volume state with oldVal
+	// if equal, replace with newVal and return true; or else, return false;
+	// waits if volume is in a state from where it can reach the ready state
+	CompareAndPutStateOrBusywait(key string, oldVal string, newVal string) bool
```

Comment:
If we pass in volume name instead of a key, this function is in fact "CompareAndPutVolStateOrBusywait".
But if you take a look at other KvStore interface methods, they are not "volume" operations. Instead, they are more generic "key-value" operations.
I know that inside the ETCD implementation we explicitly have all those volume state comparison. But still let's keep the methods at "key-value" level instead of "volume" level for consistency.